src/system_simulation/simulation_runner.rb

```diff
@@ -1,111 +1,160 @@
     sample_rate = vector.length / stop_time
 
     # https://github.com/SciRuby/rb-gsl/blob/master/examples/fft/fft.rb
     y2 = vector.fft.subvector(1, data.length - 2).to_complex2
     mag = y2.abs
-    f = GSL::Vector.linspace(0, sample_rate/2, mag.size)
-    #p mag.to_a
-    #p mag.max_index
-    #p f.to_a
-    return 1 / f[mag.max_index]
+    f = GSL::Vector.linspace(0, sample_rate / 2, mag.size)
+
+    1 / f[mag.max_index]
   end
 
-  def find_equilibrium(mass=20, constant=50)
-    # run simulation mit startTime 0.3; get first frame
-    run_simulation(constant, mass, "node_pos.*")
-    import_csv(File.join(@directory, "#{@model_name}_res.csv"))
+  # Returns index of animation frame when system is in equilibrium i.e. the mean of the angle difference
+  def find_equilibrium(constant = 50, mass = 20)
+    run_simulation(constant, mass, 'revLeft.phi')
+    raw_data = read_csv
+
+    # remove initial data point, the header
+    raw_data.shift
+    angles = raw_data.map { |data_sample| data_sample[1].to_f }
+
+    # center of oscillation
+    equilibrium_angle = angles.min + (angles.max - angles.min) / 2
+    equilibrium_data_row = raw_data.min_by do |data_row|
+      # find data row with angle that is the closest to the equilibrium
+      # (can't check for equality since we only have samples in time)
+      (equilibrium_angle - data_row[1].to_f).abs
+    end
 
-    #mocked_position_data = {"1"=>Geom::Point3d.new(0, 0, 0.748031), "2"=>Geom::Point3d.new(26.3386, 0, 0.748031), "3"=>Geom::Point3d.new(13.189, 22.8346, 0.748031), "4"=>Geom::Point3d.new(39.3307, 11.1811, 20.7874), "5"=>Geom::Point3d.new(4.48819, 12.5197, 23.4646), "6"=>Geom::Point3d.new(21.1417, -7.83465, 22.5197), "7"=>Geom::Point3d.new(22.4803, 31.378, 20.8661), "8"=>Geom::Point3d.new(42.5197, 19.8031, 0.551181), "9"=>Geom::Point3d.new(43.3376, -12.6202, 9.80675), "10"=>Geom::Point3d.new(-1.66332, 36.4207, 13.9296), "11"=>Geom::Point3d.new(57.1895, -5.2645, 31.0651), "12"=>Geom::Point3d.new(34.7027, -0.466413, 43.951), "13"=>Geom::Point3d.new(38.8624, -24.0667, 33.0068), "14"=>Geom::Point3d.new(7.85218, 47.7481, 35.3248), "15"=>Geom::Point3d.new(-10.4259, 28.9441, 37.8413), "16"=>Geom::Point3d.new(14.5224, 24.1121, 44.8706), "17"=>Geom::Point3d.new(61.3135, -28.9076, 20.0817), "18"=>Geom::Point3d.new(61.4532, -26.3095, 66.3204), "19"=>Geom::Point3d.new(-16.9368, 52.5334, 27.9541), "20"=>Geom::Point3d.new(-11.0442, 50.1143, 53.5318)}
-    #AnimationDataSample.new(0.0, mocked_position_data)
+    raw_data.index(equilibrium_data_row)
   end
 
+  def constant_for_constrained_angle(allowed_angle_delta = Math::PI / 2.0, initial_constant = 500, mass = 20,
+                                     spring_id = 0, angle_id = 0)
+    # steps which the algorithm uses to approximate the valid spring constant
+    step_sizes = [1500, 1000, 200, 50, 5]
+    constant = initial_constant
+    step_size = step_sizes.shift
+    keep_searching = true
+    abort_threshold = 50_000
+    while keep_searching
```

Comment:
I think we should somehow do another exit criterion here, as this could lead to infinite loops?

Reply:
https://github.com/robertkovax/TrussFab/pull/211/files/265a6218d9bc43494a3199d1eb5f1f1ce75e348b#diff-03ca4db6f11c8fe9f1ccf429f287c37dR105

I think this ^^^ is sufficient, since the constant will be increased continuously and eventually reach the abort threshold.